docs/ossm/ossm2-migration/network-policies.md

```diff
@@ -0,0 +1,303 @@
+- Duplicate NetworkPolicies in the same namespace should not cause issues.
+
+### Best Practices
+1. Test Network Policies in a non-production environment first.
+2. If possible, create Network Policies after migrating workloads to ensure continuous protection. If it is not possible due to security policies to migrate workloads without Network Policies, see below for tips.
+
+### Example Network Policies with both 2.6 and 3.0 present
+
+Some users may not be able to go without Network Policies during the migration process due to security concerns. 
+This can be tricky, as during the migration both control planes must have access to all workloads and vice versa. Created Network Policies must not block traffic for either control plane.
+
+To assist users in this situation, here is an example scenario with example Network Policies. 
+
+A user has a 2.6 control plane with the name `default` in the namespace `istio-system` and workloads in namespaces `httpbin-2` and `httbin-3`. They need to ensure Network Policies remain active to fulfill security obligations. 
+In order to keep Network Policies, before they disable `manageNetworkPolicy` they create Policies:
+
+Istiod Network Policy in Mesh namespace
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: istiod-default
+  namespace: istio-system
+spec:
+  ingress:
+    - {}
+  podSelector:
+    matchLabels:
+      app: istiod
+      istio.io/rev: default
+  policyTypes:
+    - Ingress
+```
+
+Expose Route Policy in Mesh namespace
+
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: expose-route-default
+  namespace: istio-system
+spec:
+  podSelector:
+    matchLabels:
+      maistra.io/expose-route: "true"
+  ingress:
+    - from:
+        - namespaceSelector:
+            matchLabels:
+              network.openshift.io/policy-group: ingress
+```
+
+Default Mesh Network Policy in Mesh namespace
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: istio-mesh-default
+  namespace: istio-system
+spec:
+  ingress:
+    - from:
+        - namespaceSelector:
+            matchLabels:
+              service-mesh: "true"
+```
+
+Expose Route Network Policy for httpbin-2 namespace
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: istio-expose-route-default
+  namespace: httpbin-2
+spec:
+  podSelector:
+    matchLabels:
+      maistra.io/expose-route: "true"
+  ingress:
+    - from:
+        - namespaceSelector:
+            matchLabels:
+              network.openshift.io/policy-group: ingress
+```
+
+Mesh Network Policy for httpbin-2 namespace
+```yaml
+apiVersion: networking.k8s.io/v1
+kind: NetworkPolicy
+metadata:
+  name: istio-mesh-default
+  namespace: httpbin-2
+spec:
+  ingress:
+    - from:
+        - namespaceSelector:
+            matchLabels:
+              service-mesh: "true"
```

Comment:
Does this mean that users will need to add this label to all of their namespaces first? Is it possible to just create a second `NetworkPolicy` identical to the ossm2 one but for the ossm3 labels?

Reply:
> Does this mean that users will need to add this label to all of their namespaces first?

Yes (or whichever label they choose)

> s it possible to just create a second NetworkPolicy identical to the ossm2 one but for the ossm3 labels?

I think I may not understand the question fully, but that sounds like what we're doing here.

Reply:
Problem here is that 2.x is using `maistra.io/member-of: istio-system` which will be gone after the migration. So new label is needed which will be there before/during/after the migration. Question is if we want to use dedicated label like `service-mesh: "true"`  or some 3.0 specific label e.g. `istio.io/rev` but that could be problem when included before the migration.
At any given time all 2.x namespaces must be able to communicate with already migrated 3.x namespaces and vice versa. Having one dedicated label seems to be the simplest to achieve that.

Reply:
>  Having one dedicated label seems to be the simplest to achieve that.

I agree on this. For naming the label, I only used `service-mesh: "true"` as an example, I think this can be left up to the users discretion?

Reply:
Trying to avoid users having to relabel all their namespaces. If you have two netpolicies, one 2.x network policy with the `maistra.io/member-of: istio-system` label and one 3.x network policy with something like:
```
ingress:
  - from:
    - namespaceSelector:
        matchLabels:
          "istio.io/rev": ""
```
The equivalent of `kubectl get networkpolicies -l "istio.io/rev" where it matches anything with the key `istio.io/rev` present.

This wouldn't work for multi-tenant but for multi-tenant they will want to use their discovery selector labels anyways.

Reply:
> service-mesh: "true"

Someone will inevitably forget the quotes and then because of yaml this will fail. Maybe `service-mesh: enabled`? Probably for the other instructions we should encourage users to do something like `maistra.io/ignore-namespace: ignore` for the same reason...

Reply:
> Someone will inevitably forget the quotes and then because of yaml this will fail. Maybe service-mesh: enabled?

Good call, made this change here. https://github.com/openshift-service-mesh/sail-operator/pull/203/commits/67917e3b764f1a3fb7da828e126f1d23eb4e7cef

I see the point on using "istio.io/rev" or maybe `istio-injection=enabled`. I guess it depends on their migration path.

Reply:
For cluster wide, we would have to cover cases when `istio.io/rev` is used vs. `istio-injection=enabled` is used and for multitenant as you mentioned, they would have to use a label used for discovery selectors. I'm not sure if we should document all those cases or maybe focus only on one which would work for all? Which is to use labels for discovery selectors and update cluster-wide doc to always use them as best practice?
Another option is to keep what we have with extra dedicated label and add a note something like: consider to reuse this label for your discovery selectors in 3.0

Reply:
Added this note https://github.com/openshift-service-mesh/sail-operator/pull/203/commits/2229bf584f46da97fef650c539563f336846275c